controllers/vspheremachine_controller.go

```diff
@@ -105,41 +114,159 @@ func AddMachineControllerToManager(ctx *context.ControllerManagerContext, mgr ma
+					oldCluster := e.ObjectOld.(*clusterv1.Cluster)
+					newCluster := e.ObjectNew.(*clusterv1.Cluster)
+					return oldCluster.Spec.Paused && !newCluster.Spec.Paused
+				},
+				CreateFunc: func(e event.CreateEvent) bool {
+					if _, ok := e.Object.GetAnnotations()[clusterv1.PausedAnnotation]; !ok {
+						return false
+					}
+					return true
+				},
+			})
+		if err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
 type machineReconciler struct {
 	*context.ControllerContext
+	VMService       services.VSphereMachineService
+	networkProvider services.NetworkProvider
+	supervisorBased bool
 }
 
+// nolint:gocognit
 // Reconcile ensures the back-end state reflects the Kubernetes resource state intent.
 func (r machineReconciler) Reconcile(ctx goctx.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	if r.supervisorBased {
+		logger := r.Logger.WithName(req.Namespace).WithName(req.Name)
+		logger.V(3).Info("Starting Reconcile VSphereMachine")
+		// Fetch the VSphereMachine instance.
+		vsphereMachine := &vmwarev1.VSphereMachine{}
+		if err := r.Client.Get(r, req.NamespacedName, vsphereMachine); err != nil {
+			if apierrors.IsNotFound(err) {
+				return reconcile.Result{}, nil
+			}
+			return reconcile.Result{}, err
+		}
+
+		// Fetch the CAPI Machine.
+		machine, err := clusterutilv1.GetOwnerMachine(r, r.Client, vsphereMachine.ObjectMeta)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		var cluster *clusterv1.Cluster
+		if machine != nil {
+			cluster = r.fetchCAPICluster(machine, vsphereMachine)
+		}
+
+		// Fetch the VSphereCluster
+		vsphereCluster := &vmwarev1.VSphereCluster{}
+		var vsphereClusterErr error
+		if cluster != nil {
+			vsphereClusterName := client.ObjectKey{
+				Namespace: vsphereMachine.Namespace,
+				Name:      cluster.Spec.InfrastructureRef.Name,
+			}
+
+			vsphereClusterErr = r.Client.Get(r, vsphereClusterName, vsphereCluster)
+		}
+		// Create the patch helper.
+		patchHelper, err := patch.NewHelper(vsphereMachine, r.Client)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrapf(
+				err,
+				"failed to init patch helper for %s %s/%s",
+				vsphereMachine.GroupVersionKind(),
+				vsphereMachine.Namespace,
+				vsphereMachine.Name)
+		}
+		// Create the machine context for this request.
+		clusterContext := &vmware.ClusterContext{
+			ControllerContext: r.ControllerContext,
+			Cluster:           cluster,
+			VSphereCluster:    vsphereCluster,
+		}
+
+		machineContext := &vmware.SupervisorMachineContext{
+			BaseMachineContext: &context.BaseMachineContext{
+				ControllerContext: r.ControllerContext,
+				Cluster:           cluster,
+				Machine:           machine,
+				Logger:            r.Logger.WithName(req.Namespace).WithName(req.Name),
+			},
+			ClusterContext: clusterContext,
+			PatchHelper:    patchHelper,
+			VSphereCluster: vsphereCluster,
+			VSphereMachine: vsphereMachine,
+		}
+		defer func() {
+			r.Recorder.EmitEvent(vsphereMachine, "Reconcile", reterr, true)
```

Comment:
I think that this can be common to both code paths

Reply:
moved into a common defer function. Couldn't keep it in a common place for both paths as defers get evaluated immediately, just not executed till function return. So had to initialize the machineContext first. If we can converge the prerequisites for reconcileDelete/Normal then we can have a common path.